lib/firenest/replicated_state.ex

```diff
@@ -0,0 +1,343 @@
+
+  It receives a list of observed remote changes for all the tracked keys.
+  A `process_state_change` specifies how the state for a single process changed.
+  In particular the change can be:
+
+    * `{initial_state, [:put, ...]}` in case the process just joined
+    * `{last_known_state, [..., :delete]}` in case the process just left
+    * `{last_known_state, [{:replace, state}, ...]}` in case it was not possible
+      to replicate incremental changes to remote state through deltas and a full
+      state transfer was performed.
+    * `{last_known_state, [..., {:delta, remote_delta}, ...]}` in case
+      incremental changes to the remote state were communicated.
+
+  This callback is optional and its behaviour depends on the value
+  of the `:remote_changes` option returned from the `c:init/2` callback.
+
+    * `:ignore` - the callback is not invoked and the server skips
+      all operations required for tracking the changes. This is the
+      default and should be chosen if precise state change tracking
+      is not required.
+
+    * `:observe_full` - calls the callback with as precise information
+      about the remote state changes as possible. This means that, for
+      example, for a very short-lived process a `[:put, :delete]` sequence
+      of changes is possible.
+
+    * `:observe_collapsed` - collapses the information about remote state
+      changes. For example a `[:put, {:delta, ...}]` sequence is collapsed
+      into just `[:put]` with all the state changes already applied and a
+      `[{:delta, ...}, :delete]` sequence is collapsed into just `[:delete]`.
+      This also means that for short-lived processes, no information
+      may be transferred if the changes would contain both `:put` and `:delete`.
+
+  This is a good place for broadcasting remote state changes.
+
+  The return value is ignored.
+  """
+  @callback observe_remote_changes(observed_remote_changes, callback_config()) :: term()
+            when observed_remote_changes: [{key(), [process_state_change]}],
+                 process_state_change: {current_state :: state(), [change]},
+                 change:
+                   :put | {:replace, new_state :: state()} | {:delta, remote_delta()} | :delete
+
+  @optional_callbacks [observe_remote_changes: 2, prepare_remote_delta: 2]
+
+  @doc """
+  Returns a child spec for the replicated state server.
+
+  Once started, each partition will call the `c:init/1` callback
+  passing all the provided options.
+
+  ## Options
+
+    * `:name` - name for the process, required;
+    * `:topology` - name of the supporting topology, required;
+    * `:partitions` - number of partitions, defaults to 1;
+    * `:broadcast_timeout` - delay of broadcasting local events to other nodes,
+      defaults to 50 ms;
+    * `:remote_changes` - specifies behaviour of the `c:observe_remote_changes/2`
+      callback and can be `:observe_full | :observe_collapsed | :ignore`,
+      defaults to `:ignore`;
```

Comment:
I thought we agreed to make this exclusively return option of the `init` callback? If a user mount something as :ignore, that should not be ignored, it won't work as expected.

Reply:
50ms seems aggressive for a default, but it might be fine. Is there prior art for a 50ms timeout?